Add RadioButton

marioharper opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description
<!--- Describe your changes in detail -->

## Related Issue
<!--- If suggesting a new feature or change, please discuss it in an issue first -->
<!--- If fixing a bug, there should be an issue describing it with steps to reproduce -->
<!--- Please link to the issue here: -->
closes #206 

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Documentation
- [ ] Bug fix (non-breaking change which fixes an issue. you didn't modify existing tests)
- [x] New feature (non-breaking change which adds functionality. you added at least one new test)
- [ ] Breaking change (fix or feature that would cause existing functionality to change. you had to modify existing tests.)
- [ ] Does this introduce a new pattern/package?

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] I have read the [**CONTRIBUTING**](https://github.com/xogroup/union/blob/master/.github/CONTRIBUTING.md) document.
- [x] I have updated the documentation as needed.
- [x] I have added tests to cover my changes.

## Reviewer Checklist

- [ ] Is change documented in a release draft? (Only if there are plans to merge PR)

marioharper commented:
@GeorgeTaveras1231 this PR is still in progress, but do you see anything in https://github.com/xogroup/union/pull/219/files#diff-65caee7278bd8b22f7c7e675638e3dac that can be optimized to use any of Unions helper classes/vars, etc?

hsbacot commented:
I think this is a solid implementation for now and this should hold unless we need more control over the radio options.  I don't think it's necessary yet but leaving an api that could work below for thought.

```jsx
<RadioGroup>
  <RadioOption value="Yes" {...otherProps} />
  <RadioOption value="No" {...otherProps} />
</RadioGroup>
```

marioharper commented:
@hsbacot I agree, I was just running with what you guys were doing on Acquisitions. I've seen it done both ways in different react radio button libraries. 

@GeorgeTaveras1231 any input on what you think would be the best API? I'm basically almost done with this PR, just finishing up a couple of unit tests.

We could run with this implementation and later implement `RadioGroup` and `RadioOption` and keep `RadioButtonList` support (which would at that point just be a different api on top of `RadioGroup` and `RadioOption`.

GeorgeTaveras1231 commented:
@marioharper I agree with you guys about `<RadioGroup>` `<RadioOption>` being a better API. This is very similar to how we have implemented the `<Dropdown>` and `<DropdownItem>`. For the sake of consistency and flexibility, I'm tempted to suggest introducing the `RadioGroup` now, as long as it isn't too much additional work. I'm concerned that if we don't introduce it now, we will live with the `RadioList` for a while; which isn't bad because honestly, it covers the most common use case.  However, it could be one of those things where a more complex use case of the radio buttons is pushed back because developers don't want to refactor the `RadioList`.

As a side note: @marioharper I don't think these `Radio*` components belong in the `tk-component-fields` package. They seem to be more thematically in line with the `fancy-fields`.

marioharper commented:
@GeorgeTaveras1231 Yeah I'll put in the work to do `RadioGroup` and `RadioOption` now. And what is "fancy-fields"? What deems something fit for fields vs fancy-fields?

GeorgeTaveras1231 commented on packages/@xo-union/tk-component-fields/src/components/RadioButtonItem/browser-spec.jsx:
```diff
@@ -0,0 +1,94 @@
+    });
+
+    it('calls handleChange when clicked', () => {
+      const clickHandlerSpy = jasmine.createSpy('clickHandlerSpy');
```
rename to `changeHandlerSpy`

GeorgeTaveras1231 commented on packages/@xo-union/tk-component-fields/src/components/RadioButtonItem/index.jsx:
```diff
@@ -0,0 +1,41 @@
+  const { label, value, name, selected, handleChange } = props;
+
+  const listItemClass = selected ?
+    RadioButtonCss.radioButtonListItemSelected : RadioButtonCss.radioButtonListItem;
```
Avoid using the camelized css modules. We are considering deprecating them.

hsbacot replied:
Any reason for deprecating them? I feel like they add value in the flow of `jsx` by providing an accessible dot property. It also allows for specific classes to be destructured from the object while importing. I'm not saying we should switch everything to it immediately, but I think it adds some value.

GeorgeTaveras1231 replied:
@hsbacot we think it adds value too which is why we enabled them in the first place. The tradeoff is that when you add a lot of these css modules in a bundle, each module is literally taking up double their size because each key is duplicated.

so which is more important? syntactic sugar or bundle size?

We still haven't deprecated them, we are waiting on #182 to be finished so that we can measure the impact of removing that setting. If it is significant enough, we will probably give up the syntactic sugar.

Another thing for us to consider is to just use camel case for all classes which will work for CSS and allow the dot syntax. The tricky part is that we inherit a lot of classes from tw/bootstrap, which means we would need a way to delete the original classes and replace them for camelized versions. Another problem is that the grid module from tw/bootstrap comes with attribute rules that depend on the dashed classes:

```css
[class^="col-"] {
  /* omitted */
}
```

hsbacot replied:
@GeorgeTaveras1231 For clarification, the duplication is coming because we are camelCasing the classes that haven't been otherwise? It's not a prevention of us using camelCased class names?

GeorgeTaveras1231 replied:
@hsbacot the duplication is coming from the generated js. The generated js when using the camelize option is:

```javascript
module.exports = {
  'a-class': 'a-class--abc123 composed-class--abc123',
  'aClass': 'a-class--abc123 composed-class--abc123',
}
```

GeorgeTaveras1231 commented on packages/@xo-union/tk-component-fields/src/components/RadioButtonItem/index.jsx:
```diff
@@ -0,0 +1,41 @@
+const RadioButtonItem = (props) => {
+  const { label, value, name, selected, handleChange } = props;
+
+  const listItemClass = selected ?
```
Should rename `selected` to `checked` to mirror the `input`'s API

GeorgeTaveras1231 commented on packages/@xo-union/tk-component-fields/src/components/RadioButtonItem/index.jsx:
```diff
@@ -0,0 +1,41 @@
+import React from 'react';
+import { PropTypes as T } from 'prop-types';
```
For consistency, use the default export and name it PropTypes

GeorgeTaveras1231 commented on packages/@xo-union/tk-component-fields/src/components/RadioButtonItem/index.jsx:
```diff
@@ -0,0 +1,41 @@
+};
+
+RadioButtonItem.propTypes = {
+  value: T.any,
```
How come the `value` prop isn't required but all other props are required?

GeorgeTaveras1231 commented on packages/@xo-union/tk-component-fields/src/css/styles/radio-button.scss:
```diff
@@ -0,0 +1,69 @@
+  margin: 0;
+  padding: 0;
+  overflow: visible;
+}
```
Use `m-0` and `p-0` from spacing.

GeorgeTaveras1231 commented on packages/@xo-union/tk-component-fields/src/css/styles/radio-button.scss:
```diff
@@ -0,0 +1,69 @@
+}
+
+.radio-button-list-item {
+  color: var-gray-4;
```
Use `composes: fg-gray-4 from '@xo-union/tk-css-colors'`

GeorgeTaveras1231 commented on packages/@xo-union/tk-component-fields/src/css/styles/radio-button.scss:
```diff
@@ -0,0 +1,69 @@
+}
+
+.radio-button-label {
+  font-size: var-fs-up-1;
```
Use the immutable `font-up-1` from typography:

```css
composes: font-up-1 from '@xo-union/tk-css-typography'`;
```

GeorgeTaveras1231 commented on packages/@xo-union/tk-component-fields/src/css/styles/radio-button.scss:
```diff
@@ -0,0 +1,69 @@
+}
+
+.radio-button-icon {
+  background-color: var-white;
```
Use `bg-white` from colors

GeorgeTaveras1231 commented on packages/@xo-union/tk-component-fields/src/css/styles/radio-button.scss:
```diff
@@ -0,0 +1,69 @@
+  height: 23px;
+  width: 23px;
+  transition: border .25s linear;
+  margin-right: var-sp-down-1;
```
Use `mr-down-1` from spacing

GeorgeTaveras1231 commented on packages/@xo-union/tk-component-fields/src/css/styles/radio-button.scss:
```diff
@@ -0,0 +1,69 @@
+  align-items: center;
+  flex-shrink: 0;
+  text-align: left;
+  margin: 0;
```
Use the immutable `m-0` from spacing
```css
composes: m-0 from '@xo-union/tk-css-spacing';
```

GeorgeTaveras1231 commented on packages/@xo-union/tk-component-fields/src/css/styles/radio-button.scss:
```diff
@@ -0,0 +1,69 @@
+  width: 23px;
+  transition: border .25s linear;
+  margin-right: var-sp-down-1;
+  margin-top: -4px;
```
Curious why you needed to use a negative margin here

GeorgeTaveras1231 commented on packages/@xo-union/tk-component-fields/src/css/styles/radio-button.scss:
```diff
@@ -0,0 +1,69 @@
+  display: flex;
+  flex-direction: row;
+  align-items: center;
+  flex-shrink: 0;
```
Curious why the `flex-shrink` is being specified without a `flex-basis` in the container.

marioharper replied:
I'm not sure, pulled this over from Acquisitions, I'm no flex master. But removing `flex-shrink` makes the radio button get all squishy when the screen size is reduced.

GeorgeTaveras1231 commented on packages/@xo-union/tk-component-fields/src/css/styles/radio-button.scss:
```diff
@@ -0,0 +1,69 @@
+  margin-top: -4px;
+  -webkit-transition: border .25s linear;
+  box-sizing: border-box;
+  flex-shrink: 0;
```
Curious why the flex-shrink is being specified without a flex-basis in the container.

GeorgeTaveras1231 commented on packages/@xo-union/tk-component-fields/src/css/styles/radio-button.scss:
```diff
@@ -0,0 +1,69 @@
+  margin-right: var-sp-down-1;
+  margin-top: -4px;
+  -webkit-transition: border .25s linear;
+  box-sizing: border-box;
```
You shouldn't need the box-sizing here. We have a global rule that applies this to all elements:

```css
html {
  box-sizing: border-box;
}

* {
  box-sizing: inherit;
}
```

GeorgeTaveras1231 commented on packages/@xo-union/tk-component-fields/src/css/styles/radio-button.scss:
```diff
@@ -0,0 +1,69 @@
+.radio-button-label {
+  font-size: var-fs-up-1;
+  cursor: pointer;
+  -webkit-transition: all 0.25s linear;
```
Could this be more specific about which properties are being transitioned?

GeorgeTaveras1231 commented on packages/@xo-union/tk-component-fields/src/css/styles/radio-button.scss:
```diff
@@ -0,0 +1,69 @@
+
+.radio-button {
+  position: absolute;
+  visibility: hidden;
```
Is the combination of these two rules the same as `display: none`?

GeorgeTaveras1231 commented:
@marioharper by `"fancy-fields"` I'm referring to the new types of fields. Also known as the Madlib style fields: https://github.com/xogroup/union/pull/202

marioharper commented:
@GeorgeTaveras1231 thanks for the feedback. Regarding the 'fancy-fields' the design team confirmed that this will be the default style of radio buttons. So I don't think this fits with fancy fields.

hsbacot commented:
I think it makes sense to refer the them as `fancy-*` because we will be moving towards northstar styles sooner rather than later.

marioharper commented:
What do you mean by northstar styles?

hsbacot commented:
The other option on the image in the issue.

On Wed, Jul 26, 2017 at 12:37 PM Mario Harper <notifications@github.com>
wrote:

> What do you mean by northstar styles?
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/xogroup/union/pull/219#issuecomment-318110604>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ABNi3nELYI1jPNuCsSIN-8IZMfepnL5zks5sR2tbgaJpZM4OivRI>
> .
>

tcmmct commented:
Based on the design unification meeting yesterday. The team aligned around these being the default radio buttons and not 'fancy.' They work for the fancy use case and the default use case

marioharper commented on packages/@xo-union/tk-component-fields/src/components/RadioButtonList/index.jsx:
```diff
@@ -0,0 +1,38 @@
+import RadioButtonItem from '../RadioButtonItem';
+
+const RadioButtonList = ({ listItems, name, selecetedValue, handleChange }) =>
+  (<ul className={RadioButtonCss.radioButtonList}>
```
what's the other option?

smanwaring replied:
You can do `{RadioButtonCss['radio-button-list']}`.

GeorgeTaveras1231 commented:
@tcmmct I'm surprised because the other fields are gray, while this radio button is blue.

Unless you guys are using the fields from #113 to determine this.

tcmmct commented:
Yeah, @GeorgeTaveras1231 we do plan to update to a style like or similar to #113.

So, The 'fancy' style will be the default. It looks slightly off from the current form fields (which is fine). It will match after we update the other form fields.

marioharper commented:
@GeorgeTaveras1231 @hsbacot this is ready for full review now. API has been changed, tests have been made, and documentation has been updated.

GeorgeTaveras1231 commented on docs/_/@components/PropTypesTable/index.jsx:
```diff
@@ -3,19 +3,31 @@ import PropTypes from 'prop-types';
+  });
+
+  return filteredObject;
+};
```
This is a candidate to be moved into `xojs` (our utility library). The library is not yet documented but it lives in `packages/xojs`. This function should be in `xojs/src/objects/omit`.

Additionally, all functions in `xojs` assume to be bound to a receiver. That just means that to be consistent, you should change `theObject` to be `this`. When used in this project, call it as:

```javascript
import omit from 'xojs/lib/objects/omit';

myObject::omit('key1', 'key2')

```

GeorgeTaveras1231 commented on packages/@xo-union/tk-component-fields/src/components/RadioGroup/index.jsx:
```diff
@@ -0,0 +1,50 @@
+  return (
+    <div className={RadioButtonCss.radioGroup}>
+      {
+        children && children.map((c, i) => (renderChild(c, i)))
```
I believe you should use `React.Children.map` here

GeorgeTaveras1231 commented on packages/@xo-union/tk-component-fields/src/components/RadioGroup/index.jsx:
```diff
@@ -0,0 +1,50 @@
+      onChange,
+      ...child.props
+    });
+  }
```
I like this approach. It seems much simpler than the implementation of `Dropdown` and `DropdownItem` which uses context.  👍

GeorgeTaveras1231 replied:
The only thing I would suggest is to move this function (`renderChild`) out of the component function. This helps avoid unnecessary memory allocation

GeorgeTaveras1231 commented on packages/@xo-union/tk-component-fields/src/css/styles/radio-button.scss:
```diff
@@ -0,0 +1,70 @@
+  mr-down-1,
+  var-sp-down-1,
+  var-sp-down-2
+) from '@xo-union/tk-css-spacing';
```
I originally preferred to use `@value` for variables only. But I like how this helps organize the file. It feels more like a js module, with all imports at the top. I approve.

GeorgeTaveras1231 commented on packages/@xo-union/tk-component-fields/src/components/RadioButton/index.jsx:
```diff
@@ -0,0 +1,44 @@
+import RadioButtonCss from '@xo-union/tk-component-fields/lib/css';
+
+const RadioButton = (props) => {
+  const { children, value, name, checked, onChange } = props;
```
Why not do the deconstructing in the arguments area? Makes sense since props is not being used for anything else.

marioharper replied:
good callout 👍

GeorgeTaveras1231 commented on docs/articles/pattern-library/core-components/form-elements/README.md:
```diff
@@ -176,3 +178,23 @@ import { Field, Dropdown, DropdownItem, Textarea, FormTheme } from '@xo-union/tk
+
+<p>{RadioGroupMeta.description}</p>
+
+<Demo cssDependencies={{ bsGrid }}>
```
The cssDependencies is not necessary here

smanwaring commented:
You can ignore the issues on Code Climate (it was turned on by mistake). We've since turned it off, but for some reason the failure from the initial check is still showing up on this PR. It should go away next time the code is updated.

marioharper commented:
@GeorgeTaveras1231 I addressed all comments except for migrating the `omit` method, I tried, but couldn't seem to get `xojs/lib/objects/omit` to resolve properly from within `docs/_/@components/PropTypesTable/index.jsx`

hsbacot commented:
This looks solid, nice work. It's a much stronger API now. 👍

GeorgeTaveras1231 commented:
@marioharper great job on this!
